### PR TITLE
fix(parser): dot property name can be escaped reserved or future reserved

### PR DESCRIPTION
**Pull request by 3cp**

### CODE REVIEW

**Comment by 3cp** [comment]
Second commit removed duplicated code, also added private identifier support in optional chaining.

File: src/parser.ts [hunk 1]
```diff
@@ -4149,7 +4148,12 @@ export function parseOptionalChain(
  * @param context Context masks
  */
 export function parsePropertyOrPrivatePropertyName(parser: ParserState, context: Context): any {
-  if ((parser.getToken() & (Token.IsIdentifier | Token.Keyword)) === 0 && parser.getToken() !== Token.PrivateField) {
+  if (
+    (parser.getToken() & (Token.IsIdentifier | Token.Keyword)) === 0 &&
+    parser.getToken() !== Token.EscapedReserved &&
+    parser.getToken() !== Token.EscapedFutureReserved &&
```
Comment (fisker) [inline src/parser.ts:4154]: Are you sure this is needed? Shouldn't the property name already allow reserved since ES5?

Comment (fisker) [reply src/parser.ts:4154]: Ah, the token is neither `Identifier` nor `Keyword`?

**Review by fisker — APPROVED** [review]

File: src/parser.ts (hunk 1 above)
Comment (3cp) [reply src/parser.ts:4154]: The normal usage was covered `x.void`, but not the escaped form. Our `EscapedReserved` and `EscapedFutureReserved` didn't carry any shared mask bit.

We should have a new mask bit called `IsIdentifierName` I think. In ECMA spec, identifier name can be anything, then identifer removed reserved words. PropertyName and ClassElementName uses IdenifierName, not Identifier.

Comment (3cp) [reply src/parser.ts:4154]: However, ESTree uses type:identifier for all cases, bit confusing.

**Pull request merged**